Implement AVX2 Gather intrinsic

fiigii opened the pull request:
Contribute to #16653

fiigii commented on src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Avx2.PlatformNotSupported.cs:
```diff
@@ -628,12 +628,12 @@ public abstract class Avx2 : Avx
         /// <summary>
         /// __m128i _mm256_i64gather_epi32 (int const* base_addr, __m256i vindex, const int scale)
-        ///   VPGATHERQD ymm, vm64y, ymm
+        ///   VPGATHERQD xmm, vm64y, xmm
```
Fixed some doc mistake of Gather intrinsic

fiigii commented on src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Avx2.cs:
```diff
@@ -546,243 +546,963 @@ public abstract class Avx2 : Avx
+                case 8:
+                    return GatherVector128(baseAddress, index, 8);
+                default:
+                    throw new ArgumentOutOfRangeException();
```
The `scale` parameter has discrete valid values, so the RyuJIT bound-check does not work. Use managed non-const fallback to handle Gather intrinsic, which we do need to worry about compiler optimization on the switch table.

fiigii replied:
Added a comment "The scale parameter should be 1, 2, 4 or 8, otherwise, ArgumentOutOfRangeException will be thrown." for each function header.

fiigii commented on src/jit/gentree.h:
```diff
@@ -4112,6 +4112,7 @@ struct GenTreeSIMD : public GenTreeJitIntrinsic
 struct GenTreeHWIntrinsic : public GenTreeJitIntrinsic
 {
     NamedIntrinsic gtHWIntrinsicId;
+    var_types gtIndexBaseType; // for AVX2 Gather* intrinsics
```
Gather intrinsics have complex overloads that need additional information (the base-type of index vector) for codegen, so adding a field in IR. But that let `GenTreeHWIntrinsic` become a large node. @CarolEidt do you think it ok?

fiigii commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -237,10 +236,7 @@ int HWIntrinsicInfo::lookupNumArgs(const GenTreeHWIntrinsic* node)
-#endif
-
-        return 3;
+        return numArgs;
```
`GatherMaskVector*` have 5 parameters that breaks our previous `numArgs` assumption.

tannergooding replied:
Do we actually have to carry all 5 args through? They eventually get folded down to a `target register`, `vm32x`, and `mask register`, so I would think we we could do some of this folding in the importer.

fiigii replied:
> so I would think we we could do some of this folding in the importer.

That seems to require a new IR to contain the folded VSIB.

tannergooding replied:
Yeah, I guess we could only fold it to 3 if we knew the register for `index` before-hand....

It just seems bad to make this many changes for a single family of instructions, since it will impact the 90% case (which is the other HWIntrinsics)...

tannergooding commented:
This section is important and we should ensure it doesn't impact any existing assumptions:
![image](https://user-images.githubusercontent.com/10487869/43929418-bbff6c90-9be9-11e8-9e91-8588fc516cc8.png)

fiigii commented:
> This section is important and we should ensure it doesn't impact any existing assumptions:

Yes, I have addressed all the special encoding requests. But some fault issues (like the last one in the above list) are difficult to test...

4creators commented on src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Avx2.cs:
```diff
@@ -546,243 +546,963 @@ public abstract class Avx2 : Avx
         /// </summary>
-        public static unsafe Vector128<int> GatherVector128(int* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
+        public static unsafe Vector128<int> GatherVector128(int* baseAddress, Vector128<int> index, byte scale)
+        {
```
Perhaps we should apply `AggressiveInlining` attribute to all managed implementations

fiigii replied:
No, the fallback should always be a call-node.

4creators commented on src/jit/emitxarch.cpp:
```diff
@@ -4326,6 +4330,53 @@ void emitter::emitIns_R_R_AR(instruction ins, emitAttr attr, regNumber reg1, reg
+                             regNumber   index,
+                             int         scale,
+                             int         offs)
+{
```
Coding convention: method needs header

fiigii replied:
Thanks

fiigii replied:
Done

4creators commented on src/jit/emitxarch.cpp:
```diff
@@ -4326,6 +4330,53 @@ void emitter::emitIns_R_R_AR(instruction ins, emitAttr attr, regNumber reg1, reg
     emitCurIGsize += sz;
 }
 
+bool IsAVX2GatherInstruction(instruction ins)
```
Coding convention: method needs header

fiigii replied:
Will do.

fiigii commented:
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic

CarolEidt commented on src/jit/emitxarch.cpp:
```diff
@@ -4326,6 +4330,67 @@ void emitter::emitIns_R_R_AR(instruction ins, emitAttr attr, regNumber reg1, reg
     emitCurIGsize += sz;
 }
 
+// return true if ins is an AVX2 Gather instruction
```
Since this is new, you should add a header here as well.

fiigii replied:
Will do, thanks.

CarolEidt commented on src/jit/emitxarch.h:
```diff
@@ -333,6 +333,15 @@ void emitIns_R_R_A(instruction ins, emitAttr attr, regNumber reg1, regNumber reg
+                    regNumber   base,
+                    regNumber   index,
+                    int         scale,
+                    int         offs);
```
I don't know if this was formatted this way by jit-format, but I'd prefer to see it declared in a format more similar to those around it.

fiigii replied:
Yes, this was given by jit-format, let me try to unify the format.

fiigii replied:
This format is given by clang-format and cannot manually change...

CarolEidt commented on src/jit/gentree.cpp:
```diff
@@ -6867,23 +6868,23 @@ void Compiler::gtBlockOpInit(GenTree* result, GenTree* dst, GenTree* srcOrFillVa
+     * isn't handling self-assignment of struct variables correctly.  This issue may not
+     * surface if struct promotion is ON (which is the case on x86/arm).  But still the
+     * fundamental issue exists that needs to be addressed.
+     */
```
Why was this changed?

CarolEidt commented on src/jit/gentree.h:
```diff
@@ -479,8 +479,8 @@ struct GenTree
-        INDEBUG(gtCostsInitialized =
-                    tree->gtCostsInitialized;) // If the 'tree' costs aren't initialized, we'll hit an assert below.
+        INDEBUG(gtCostsInitialized = tree->gtCostsInitialized;) // If the 'tree' costs aren't initialized, we'll hit an
+                                                                // assert below.
```
Again, not sure why this changed - but if it's going to change, the comment should be moved above.

fiigii replied:
Hmm, on my computer, jit-format + vs code is not stable on a few jit source files, which can output different format with each run, especially for some comments or whitespace. 
Will fix, sorry for the mess.

CarolEidt commented on src/jit/gentree.h:
```diff
@@ -4112,6 +4112,7 @@ struct GenTreeSIMD : public GenTreeJitIntrinsic
 struct GenTreeHWIntrinsic : public GenTreeJitIntrinsic
 {
     NamedIntrinsic gtHWIntrinsicId;
+    var_types      gtIndexBaseType; // for AVX2 Gather* intrinsics
```
It is probably not a big issue for intrinsics, though for methods with heavy intrinsic usage it could be an impact. Did you consider deriving from `GenTreeHWIntrinsic` to make a specialized node?

CarolEidt replied:
It might also be worth considering whether to define something like `GenTreeHWIntrinsicBig`, to isolate both the additional fields as well as the extra operands.

fiigii replied:
> Perhaps make NamedIntrinsic unsigned short? 64k intrinsics ought to be enough for anybody.

@mikedn Good point! Yes, 64k is definitely enough for the foreseeable future (AVX-512 based ISAs).

fiigii replied:
> make NamedIntrinsic unsigned short

Made this change, now `GenTreeHWIntrinsic` is still a small node.

tannergooding replied:
What is the `gtIndexBaseType` actually needed for? Can it not be inferred from the rest of the signature?

CarolEidt commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -2002,6 +2004,115 @@ void CodeGen::genAvxOrAvx2Intrinsic(GenTreeHWIntrinsic* node)
+                addrIndexReg = op3Reg;
+                indexOp      = op3;
+
+                // the mask register will be cleaned by gather instructions
```
will be cleared by ...?

fiigii replied:
Oops, will fix.

CarolEidt commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -1177,6 +1177,8 @@ void CodeGen::genHWIntrinsicJumpTableFallback(NamedIntrinsic            intrinsi
                                               HWIntrinsicSwitchCaseBody emitSwCase)
 {
     assert(nonConstImmReg != REG_NA);
+    // AVX2 Gather intrinsics use managed non-const fallback
```
This deserves more explanation.

fiigii replied:
Thanks, will add more comments.

CarolEidt commented on src/jit/lowerxarch.cpp:
```diff
@@ -2576,6 +2576,17 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
+
     if (!HWIntrinsicInfo::SupportsContainment(intrinsicId))
     {
         // Exit early if containment isn't supported
```
You're no longer really exiting early - was there some reason you moved the immediate containment above this?

fiigii replied:
Before this PR, IMM intrinsics do not have NoContainment ones. But AVX2 `Gather*` are IMM intrinsic AND memory-load intrinsic that cannot work with containment (so they have NoContainment flag).

So we need to moving the immediate containment above the early exiting, which marks the IMM argument of Gather before return.

fiigii replied:
Before this PR, IMM intrinsics do not have NoContainment ones. But AVX2 `Gather*` are IMM intrinsic AND memory-load intrinsic that cannot work with containment (so they have NoContainment flag).

So we need to moving the immediate containment above the early exiting, which marks the IMM argument of Gather before return.

CarolEidt replied:
So then I think this comment needs to be deleted.

fiigii commented:
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic

fiigii commented:
Addressed feedback @tannergooding @CarolEidt PTAL

fiigii commented:
Addressed feedback @tannergooding @CarolEidt PTAL

fiigii commented:
Addressed feedback. @tannergooding @CarolEidt could you please take a look? Some applications https://github.com/dotnet/machinelearning/pull/691 need the Gather intrinsic.

fiigii commented:
@tannergooding @CarolEidt ping?

CarolEidt commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -2008,6 +2011,115 @@ void CodeGen::genAvxOrAvx2Intrinsic(GenTreeHWIntrinsic* node)
+                addrIndexReg = op3Reg;
+                indexOp      = op3;
+
+                // the mask register will be cleared by gather instructions
```
I'm not sure I understand the point of this comment. The `maskReg` is a tempreg, so I don't see that it matters whether it will be cleared. Perhaps you mean that we need to copy op4 to a temp reg to use for the mask because it will be over-written?

fiigii replied:
> Perhaps you mean that we need to copy op4 to a temp reg to use for the mask because it will be over-written?

Yes, will update the comment.

CarolEidt commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -2008,6 +2011,115 @@ void CodeGen::genAvxOrAvx2Intrinsic(GenTreeHWIntrinsic* node)
+                if (targetReg != op1Reg)
+                {
+                    // copy source vector to the target register for masking merge
+                    emit->emitIns_R_R(INS_movaps, attr, targetReg, op1Reg);
```
What if `op2Reg == targetReg`? I don't think it's been set as delayFree.

fiigii replied:
Op2Reg is a GPR (base register of the address) and cannot be same as the vector target register.

CarolEidt replied:
Right - thanks for clarifying.

CarolEidt commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -2008,6 +2011,115 @@ void CodeGen::genAvxOrAvx2Intrinsic(GenTreeHWIntrinsic* node)
+            }
+
+            regNumber lastOpReg = lastOp->gtRegNum;
+            genConsumeRegs(lastOp);
```
What is the purpose of having this here, rather than always consuming op3 before the if-stmt above, and then taking care of the other sources in that if stmt?

CarolEidt commented on src/jit/lsraxarch.cpp:
```diff
@@ -2581,6 +2588,44 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
+            case NI_AVX2_GatherMaskVector256:
+            {
+                assert(numArgs == 5);
+                // Any pair of the index, mask, or destination registers should be different
```
This comment mentions index, mask and destination; it needs to map those to the opN names. Also, what about the address? Doesn't that also need to be distinct from the target, as the target will be overwritten?

fiigii replied:
Thanks, will update the comment.

> Also, what about the address? Doesn't that also need to be distinct from the target, as the target will be overwritten?

Don't you mean the address base register? If yes, that is a GPR and cannot be same as the vector target register.

fiigii commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -532,7 +592,8 @@ GenTree* Compiler::addRangeCheckIfNeeded(NamedIntrinsic intrinsic, GenTree* last
     assert(lastOp != nullptr);
     // Full-range imm-intrinsics do not need the range-check
     // because the imm-parameter of the intrinsic method is a byte.
```
Done.

fiigii commented:
@CarolEidt Thank you for the review and suggestions, I have addressed all the feedback.

CarolEidt commented:
@tannergooding - did you want to review this before merging?

fiigii commented:
@tannergooding ping?

tannergooding commented on src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Avx2.cs:
```diff
@@ -545,244 +545,1012 @@ public abstract class Avx2 : Avx
+                case 8:
+                    return GatherVector128(baseAddress, index, 8);
+                default:
+                    throw new ArgumentOutOfRangeException();
```
nit: Should be `throw new ArgumentOutOfRangeException(nameof(scale))`

tannergooding replied:
(same elsewhere as well)

fiigii replied:
Other intrinsic throw the exception without the argument name. Shall we keep consistent?

fiigii replied:
Agree, logged at https://github.com/dotnet/coreclr/issues/19689

tannergooding commented on src/jit/emitxarch.cpp:
```diff
@@ -4365,6 +4369,74 @@ void emitter::emitIns_R_R_AR(instruction ins, emitAttr attr, regNumber reg1, reg
+//    scale - the scale number of VSIB
+//    offs - the offset added to the memory address from base
+//
+void emitter::emitIns_R_AR_R(instruction ins,
```
I wonder if we shouldn't be encoding the scale/offs in an `Indir` node and letting `emitHandleMemOp` handle it

@CarolEidt ?

CarolEidt replied:
The scale/offs would be on a `GenTreeAddrMode` - presumably a `GT_LEA`. It would make sense for that to be a child of the gather node, which would not only consolidate the scale/offs handling, but also would reduce the operand count of the gather.

fiigii replied:
@CarolEidt Thanks for the comments. Do you suggest to make the change in this PR? Or I can improve it in a new PR?

CarolEidt replied:
I think it would be fine (and probably a bit cleaner) to do it as a separate PR.

tannergooding commented on src/jit/gentree.cpp:
```diff
@@ -17522,6 +17522,11 @@ bool GenTreeHWIntrinsic::OperIsMemoryLoad()
     {
         // Some AVX instructions here also have MemoryLoad sematics
 
+        if (HWIntrinsicInfo::isAVX2GatherIntrinsic(gtHWIntrinsicId))
```
Why was this inserted here, rather than as part of the rest of the logic flow below?

tannergooding replied:
The check would have remained correct and simple by changing the code as such:
```diff
- // Do we have 3 operands?
- if (HWIntrinsicInfo::lookupNumArgs(this) != 3)
+ // Do we have less than 3 operands?
+ if (HWIntrinsicInfo::lookupNumArgs(this) < 3)
```

This also ensures that we aren't adding a "fast-path" for checks that don't need it

fiigii replied:
Good point, will do.

tannergooding commented on src/jit/gentree.h:
```diff
@@ -479,8 +479,8 @@ struct GenTree
     {
-        INDEBUG(gtCostsInitialized =
-                    tree->gtCostsInitialized;) // If the 'tree' costs aren't initialized, we'll hit an assert below.
+        // If the 'tree' costs aren't initialized, we'll hit an assert below.
```
Why are all these random comment re-alignments popping up? They seem unrelated to the PR.

tannergooding commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -2008,6 +2011,113 @@ void CodeGen::genAvxOrAvx2Intrinsic(GenTreeHWIntrinsic* node)
+            genConsumeRegs(op2);
+            genConsumeRegs(op3);
+
+            if (intrinsicId == NI_AVX2_GatherMaskVector128 || intrinsicId == NI_AVX2_GatherMaskVector256)
```
I would think a comment or an assert would be sufficient for that

fiigii replied:
Ok, will do.

tannergooding commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -532,7 +592,10 @@ GenTree* Compiler::addRangeCheckIfNeeded(NamedIntrinsic intrinsic, GenTree* last
-    if (mustExpand && !HWIntrinsicInfo::HasFullRangeImm(intrinsic) && HWIntrinsicInfo::isImmOp(intrinsic, lastOp))
+    // AVX2 Gather intrinsics no not need the range-check
+    // because their imm-parameter have discrete valid values that are handle by managed code
+    if (mustExpand && !HWIntrinsicInfo::isAVX2GatherIntrinsic(intrinsic) &&
```
It seems like this should be the last check in the `&&` chain, since the other two are more likely to exit the chain first

fiigii replied:
Agree, will fix

tannergooding commented on src/jit/lowerxarch.cpp:
```diff
@@ -2576,9 +2576,19 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
     GenTree* op2 = node->gtGetOp2();
     GenTree* op3 = nullptr;
 
+    if (HWIntrinsicInfo::lookupCategory(intrinsicId) == HW_Category_IMM)
```
Why was this moved up?

fiigii replied:
Before this PR, IMM intrinsics do not have NoContainment ones. But AVX2 Gather* are IMM intrinsic AND memory-load intrinsic that cannot work with containment (so they have NoContainment flag).

So we need to move the immediate containment above the early exiting, which marks the IMM argument of Gather before return.

tannergooding replied:
Still not sure I see the problem.

There are multiple intrinsics which can be sometimes contained, or partially contained (just the immediate node). You don't mark those as `NoContainment`, but exit below instead.

tannergooding commented on tests/src/JIT/HardwareIntrinsics/X86/Avx2/GatherMaskVector128_r.csproj:
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Please add these as templated tests, since we are trying to reduce the number of manually managed tests that we have for the intrinsics.

tannergooding replied:
and when doing so, please ensure that the new templates follow the same general pattern as previous templates, to ensure that all the special-cases are covered (such as indirect invocation)

fiigii replied:
Gather intrinsic have very special overloads and semantics that do not work with the current test template framework.

Let me move the test cases later in a new PR, some applications are waiting for these intrinsic

tannergooding replied:
Then, at the very least, you need to ensure all the special cases are still covered in the manual code.

The current templates cover 14 or so scenarios that are meant to test the various code patterns that must be supported and that might influence the JIT (including things like `indirect invocation`, various patterns that may impact `containment`, etc).

tannergooding replied:
I'm also not sure I agree with "does not work with the test template framework".

The framework is very generic and just works with a template and named string values. You should be able to create a template that fits your needs (and that follows the existing pattern) and have it just work.

fiigii commented:
@tannergooding @CarolEidt Thank you for the review. Addressed all the feedback.

Will move the tests to template and investigate the indir solution later.

tannergooding commented:
I would still like to see test coverage for the indirect invocation case, to validate that it is working as intended.

fiigii commented:
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic

fiigii commented:
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd

fiigii commented:
> I would still like to see test coverage for the indirect invocation case,

@tannergooding I have added indirect invocation cases, can we merge this PR?

fiigii commented:
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd

tannergooding commented:
I'm fine with merging. I would appreciate it if you could log a bug tracking the test code cleanup (porting it to a template, etc).

@CarolEidt, any other feedback on your end?

fiigii commented:
@tannergooding thanks, will do.

fiigii commented:
Logged at https://github.com/dotnet/coreclr/issues/19825

CarolEidt commented:
No additional feedback - I'm happy with merging

tannergooding commented:
Will merge after the last test finishes (it had timed out previously, so I restarted it).

fiigii commented:
@tannergooding could you please merge this PR?

fiigii commented:
@tannergooding could you please merge this PR?

tannergooding commented:
Done. Thanks for the work @fiigii
